File: csharp/783.minimum-distance-between-bst-nodes.cs
/*
 * @lc app=leetcode id=783 lang=csharp
 *
 * [783] Minimum Distance Between BST Nodes
 */

using System;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    private static bool MinDiffInBSTRecurse(TreeNode root, out int diff, out int min, out int max) {
        if (root == null) {
            diff = -1;
            min = int.MaxValue;
            max = int.MinValue;
            return false;
        }
        diff = int.MaxValue;
        min = root.val;
        max = min;
        if (MinDiffInBSTRecurse(root.left, out int ldiff, out int lmin, out int lmax)) {
            diff = Math.Min(ldiff, root.val - lmax);
            min = lmin;
        }
        if (MinDiffInBSTRecurse(root.right, out int rdiff, out int rmin, out int rmax)) {
            diff = Math.Min(diff, Math.Min(rdiff, rmin - root.val));
            max = rmax;
        }
        return true;
    }

    public int MinDiffInBST(TreeNode root) {
        if (MinDiffInBSTRecurse(root, out int diff, out int min, out int max)) {
            return diff;
        }
        return -1;
    }
}
// @lc code=end

File: java/783.minimum-distance-between-bst-nodes.java
/*
 * @lc app=leetcode id=783 lang=java
 *
 * [783] Minimum Distance Between BST Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private static void addNodes(List<Integer> ret, TreeNode root) {
        if (root == null) {
            return;
        }
        addNodes(ret, root.left);
        ret.add(root.val);
        addNodes(ret, root.right);
    }

    public int minDiffInBST(TreeNode root) {
        List<Integer> values = new ArrayList<Integer>();
        addNodes(values, root);
        int n = values.size(), ret = Integer.MAX_VALUE;
        for (int i = 0; i < n - 1; ++i) {
            ret = Math.min(ret, values.get(i + 1) - values.get(i));
        }
        return ret;
    }
}
// @lc code=end

